### DISCUSSION TITLE
[PLANNING] VISION

### DISCUSSION THREAD

**Discussion by jkowalleck** [CycloneDX/cyclonedx-node-yarn#8] [Ideas]
the users of the old <https://github.com/CycloneDX/cyclonedx-node-module/tree/3.x>
expected support for yarn, 
but the yarn support was limited and flawed. see the closed-issues list of the project ... 

to bring yarn support to the community, this dedicated project should solve the feature lack.

### Responsibilities

* [x] create SBOM in JSON or XML from yarn environments
* [ ] ... to be continued ...

### Capabilities / Features

- [x] runs in CLI
- [x] can create reproducible results (feature switch)
- [x] can create results in CDX spec 1.2 to 1.x (feature switch)
- [x] can create results in XML or JSON  (feature switch)
- [x] supports latest _yarn_ version: v4  
- [ ] support _yarn_ v1, v2, v3  
      ~~if this is not possible, the project could split for dedicated support of the versions~~
      became optional; like nice to have; like if a low-hanging fruit, why not 
- [x] MUST NOT publish any API or internal interfaces. 
- [ ] ... to be continued ...

## Architecture 

this project should utilize https://github.com/CycloneDX/cyclonedx-javascript-library 

### Option: yarn plugin 

implement a yarn plugin, that adds a new yarn command an can produce the SBOM as expected.
https://yarnpkg.com/features/plugins

need to do a Prove of concept, that showed that the plugin can run a command, and this command has access to yarns internal data structures ... 
or find other capabilities ....

## misc 

* ... to be continued ...

----

You want to contribute or champion this tool? ❤️  see https://github.com/CycloneDX/cyclonedx-node-yarn/issues/12

**Comment by jkowalleck** [comment]
the implementation is coming to an end, nearly all core features are implemented. MVP in reach.
discussion closed.

**Comment by jdalton** [comment]
Which version of yarn does this support (and versions of the lock file)?

**Reply by jkowalleck** [reply]
> Which version of yarn does this support [...]?
currently, there is a plugin for yarn v4

see https://github.com/CycloneDX/cyclonedx-node-yarn/blob/2c87fcc6922bbe66192fd99d679676456aad0bcf/README.md?plain=1#L29-L35
see https://github.com/CycloneDX/cyclonedx-node-yarn/blob/2c87fcc6922bbe66192fd99d679676456aad0bcf/package.json#L88-L91

> [...] versions of the lock file?

this is a yarn plugin. it supports everything that is supported by the yarn version you use/provide.

**Reply by jkowalleck** [reply]
i am currently investigating support foe yarn3, alongside with support for yarn4. 
looks promising :D see https://github.com/CycloneDX/cyclonedx-node-yarn/pull/121

**Reply by jkowalleck** [reply]
Support for yarn3 was added in [v1.0.0-rc.2](https://github.com/CycloneDX/cyclonedx-node-yarn/releases/tag/v1.0.0-rc.2)
